### PR TITLE
RTPS contrib layer

### PR DESCRIPTION
**Pull request by phretor**
Check this branch for detailed history of this change: https://github.com/phretor/scapy/commits/rtps

**Disclaimer:** I'm not sure this is the right way to bind layers that are not restricted to a non-fixed port range, so I put a comment. Let me or @vmayoral know if you prefer to enable the "ugly hack". All dissection tests pass even now.

```
# layer binding
# for port in range(7400, 7500):  # very ugly hack :-)
#     bind_layers(UDP, RTPS, dport=port)
#     bind_layers(TCP, RTPS, dport=port)

bind_layers(UDP, RTPS)
bind_layers(RTPS, RTPSMessage, magic=b"RTPS")
bind_layers(RTPS, RTPSMessage, magic=b"RTPX")
```

Co-authored-by: Federico Maggi <fede@maggi.cc>
Co-authored-by:  Víctor Mayoral Vilches <v.mayoralv@gmail.com>
Signed-off-by: Federico Maggi <fede@maggi.cc>

### CODE REVIEW

**Comment by phretor** [comment]
Looks like a test unrelated to this PR is not passing.

```
######
## ISOTPSoftSocket MITM attack tests
######
###(047)=[failed] Send a two-frame ISOTP message with padding
>>> acks = TestSocket(CAN)
>>> cans = TestSocket(CAN)
>>> acks.pair(cans)
>>> 
>>> acker = AsyncSniffer(opened_socket=acks, store=False, prn=lambda x: acks.send(CAN(identifier = 0x241, data=dhex("30 00 00"))), timeout=1, count=1)
>>> acker.start()
>>> with TestSocket(CAN) as isocan, ISOTPSoftSocket(isocan, sid=0x641, did=0x241, padding=True) as s:
...     acks.pair(isocan)
...     cans.pair(isocan)
...     s.send(ISOTP(data=dhex("01 02 03 04 05 06 07 08")))
...     can = cans.sniff(timeout=1, count=1)[0]
...     assert(can.identifier == 0x641)
...     assert(can.data == dhex("10 08 01 02 03 04 05 06"))
...     can = cans.sniff(timeout=1, count=1)[0]
...     assert(can.identifier == 0x241)
...     assert(can.data == dhex("30 00 00"))
...     can = cans.sniff(timeout=1, count=1)[0]
...     assert(can.identifier == 0x641)
...     assert(can.data == dhex("21 07 08 CC CC CC CC CC"))
... 
11
Traceback (most recent call last):
  File "<input>", line 7, in <module>
AssertionError
UTscapy ended with error code 1
SocketCAN support: False
Using PythonCANSocket virtual
SocketCAN support: False
Using PythonCANSocket virtual
SocketCAN support: False
Using PythonCANSocket virtual
SocketCAN support: False
Using PythonCANSocket virtual
SocketCAN support: False
Using PythonCANSocket virtual
SocketCAN support: False
Using PythonCANSocket virtual
SocketCAN support: False
Using PythonCANSocket virtual
SocketCAN support: False
Using PythonCANSocket virtual
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
TEST SKIPPED: can-isotp not available
SocketCAN support: False
Using PythonCANSocket virtual
ERROR: InvocationError for command 'C:\projects\scapy\.tox\py37-windows\Scripts\python' -m coverage run -m scapy.tools.UTscapy -c test/configs/windows.utsc (exited with code 1)
___________________________________ summary ___________________________________
ERROR:   py37-windows: commands failed
Command exited with code 1
```

**Comment by phretor** [comment]
Bump ☺️

**Comment by polybassa** [comment]
Thanks. I'll inspect those failures.
-- 
Viele Grüße

Nils Weiß

File: scapy/contrib/rtps/types.py
Comment (gpotter2) [reply scapy/contrib/rtps/types.py:147]: I think you should remove it (or if you really want to keep it, move it to `scapy/tools/rtps/....py`)

File: test/contrib/rtps/rtps.uts
Comment (gpotter2) [inline test/contrib/rtps/rtps.uts:47]: Could you include some slightly better build tests?
A single build test with only the default value is a bit light. At least try to add a few submessages . The dissection test is fine though.

**Comment by gpotter2** [comment]
Also please rebase against master

File: scapy/contrib/rtps/types.py
Comment (phretor) [reply scapy/contrib/rtps/types.py:147]: Agreed. I'll remove it.

File: test/contrib/rtps/rtps.uts
Comment (phretor) [reply test/contrib/rtps/rtps.uts:47]: I think the following is fine? `p = RTPS(pkt)`

**Comment by phretor** [comment]
> Also please rebase against master

See the most recent commits. I rebased `rtps` against `master`. If you can get this merged quickly, I know @vmayoral has some improvements ready for a PR ;-)

**Comment by vmayoral** [comment]
> See the most recent commits. I rebased `rtps` against `master`. If you can get this merged quickly, I know @vmayoral has some improvements ready for a PR ;-)

+1, it'd be awesome to get this in ASAP folks.

Count on us for maintaining this down the road (i.e. working on some extensions at https://github.com/vmayoral/scapy/tree/rtps). We'll push those once this first contribution is finalized.

**Comment by polybassa** [comment]
Hi, I think you made a mistake in your rebase/merge. You included several already merged commits in this PR. Could you please exclude them.

**Comment by phretor** [comment]
> Hi, I think you made a mistake in your rebase/merge. You included several already merged commits in this PR. Could you please exclude them.

Yes. Rebased is to be used with care and I've been careless lately :-)

**Comment by phretor** [comment]
Should be good now right?

**Comment by guedou** [comment]
There are still several issues:
- failing flake8 tests: https://github.com/secdev/scapy/runs/4125077297?check_suite_focus=true#step:5:12
- missing contrib infos in rtsp.py: https://github.com/secdev/scapy/runs/4125077368?check_suite_focus=true#step:5:1645

**Comment by phretor** [comment]
> There are still several issues:
> 
>     * failing flake8 tests: https://github.com/secdev/scapy/runs/4125077297?check_suite_focus=true#step:5:12
> 
>     * missing contrib infos in rtsp.py: https://github.com/secdev/scapy/runs/4125077368?check_suite_focus=true#step:5:1645

I'm having a hard time running test cases locally, which in turns prevents me from uploading sane code. For example, this failure occurs every time, causing the important test failures to be buried into miles of screen output:

```
UTscapy ended with error code 1
ERROR: InvocationError for command /Users/federico_maggi/Projects/Prj-DDS/dds-scapy/scapy/.tox/py27-bsd_non_root/bin/python -m coverage run -m scapy.tools.UTscapy -c test/configs/bsd.utsc -K manufdb -K tshark -N (exited with code 1)
```

Flake8 checks are easy, and that's my mistake having left those behind, but I've never managed to run the full `tox` cases locally without failures. Any ideas why? Any way I can improve my local dev workflow?

**Comment by gpotter2** [comment]
You should be able to run `tox -e flake8,mypy`

**Comment by phretor** [comment]
> You should be able to run `tox -e flake8,mypy`

Looks like other layers could use some polishing, too, but this one should be good to go now.

```
flake8 installed: flake8==4.0.1,mccabe==0.6.1,pycodestyle==2.8.0,pyflakes==2.4.0
flake8 run-test-pre: PYTHONHASHSEED='3378550810'
flake8 run-test: commands[0] | flake8 scapy/
mypy installed: mypy==0.790,mypy-extensions==0.4.3,typed-ast==1.4.3,typing==3.7.4.3,typing-extensions==3.10.0.2
mypy run-test-pre: PYTHONHASHSEED='3378550810'
mypy run-test: commands[0] | python .config/mypy/mypy_check.py
scapy/utils.py:1059: error: unused 'type: ignore' comment
scapy/utils.py:1075: error: unused 'type: ignore' comment
scapy/utils.py:2365: error: Module 'ctypes' has no attribute 'windll'
scapy/supersocket.py:299: error: Module has no attribute "AF_PACKET"
scapy/supersocket.py:323: error: Module has no attribute "PACKET_OUTGOING"
scapy/automaton.py:99: error: Module has no attribute "windll"
scapy/automaton.py:107: error: Module has no attribute "windll"
scapy/automaton.py:120: error: Module has no attribute "windll"
scapy/automaton.py:146: error: Module has no attribute "windll"
scapy/automaton.py:153: error: Module has no attribute "windll"
scapy/automaton.py:155: error: Module has no attribute "FormatError"
scapy/automaton.py:159: error: Module has no attribute "windll"
scapy/automaton.py:161: error: Module has no attribute "FormatError"
scapy/automaton.py:165: error: Module has no attribute "windll"
scapy/automaton.py:167: error: Module has no attribute "FormatError"
scapy/automaton.py:173: error: Returning Any from function declared to return "int"
scapy/arch/linux.py:213: error: unused 'type: ignore' comment
scapy/arch/linux.py:491: error: Module has no attribute "AF_PACKET"
scapy/arch/linux.py:564: error: Module has no attribute "PACKET_OUTGOING"
scapy/contrib/cansocket_native.py:62: error: Module has no attribute "PF_CAN"
scapy/contrib/cansocket_native.py:64: error: Module has no attribute "CAN_RAW"
scapy/contrib/cansocket_native.py:66: error: Module has no attribute "SOL_CAN_RAW"
scapy/contrib/cansocket_native.py:67: error: Module has no attribute "CAN_RAW_RECV_OWN_MSGS"
scapy/contrib/cansocket_native.py:86: error: Module has no attribute "SOL_CAN_RAW"
scapy/contrib/cansocket_native.py:87: error: Module has no attribute "CAN_RAW_FILTER"
scapy/contrib/isotp/isotp_native_socket.py:213: error: Module has no attribute "CAN_EFF_FLAG"
scapy/contrib/isotp/isotp_native_socket.py:215: error: Module has no attribute "CAN_EFF_FLAG"
scapy/contrib/isotp/isotp_native_socket.py:218: error: Module has no attribute "PF_CAN"
scapy/contrib/isotp/isotp_native_socket.py:292: error: Module has no attribute "PF_CAN"
Found 29 errors in 6 files (checked 64 source files)
ERROR: InvocationError for command /Users/federico_maggi/Projects/Prj-DDS/dds-scapy/scapy/.tox/mypy/bin/python .config/mypy/mypy_check.py (exited with code 1)
__________________________________________________ summary __________________________________________________
  flake8: commands succeeded
ERROR:   mypy: commands failed
>>> elapsed time 41s
```

File: scapy/contrib/rtps/common_types.py
Comment (guedou) [inline scapy/contrib/rtps/common_types.py:200]: This shoud be ByteField.__init__ (see https://app.codacy.com/gh/secdev/scapy/pullRequest?prid=8385225)

File: scapy/contrib/rtps/common_types.py
Comment (guedou) [inline scapy/contrib/rtps/common_types.py:153]: It think that it is better to replace this assert by a warning and a default value.

**Review by guedou — CHANGES_REQUESTED** [review]
This PR looks good. I made two, likely final comments. The AppVeyor failures are not related to this PR.

My apologies for not having the time to review this before BHEU. I know that this is an important deadline for you =/

File: scapy/contrib/rtps/common_types.py
Comment (gpotter2) [reply scapy/contrib/rtps/common_types.py:200]: They should probably extend `LEIntField` if the default format is `<I`

File: scapy/contrib/rtps/common_types.py
Comment (phretor) [reply scapy/contrib/rtps/common_types.py:153]: I'm about to commit this:

```python
    def set_endianness(self, pkt):
        if getattr(pkt, "endianness", None) is not None:
            self.endianness = pkt.endianness
        elif self.endianness_from is not None and pkt:
            self.endianness = self.endianness_from(pkt)
            if self.endianness is None:
                warnings.warn(
                    'Endianess should never be None.'
                    'Setting it to default: {}', DEFAULT_ENDIANESS)
                self.endianness = DEFAULT_ENDIANESS
```

File: scapy/contrib/rtps/common_types.py
Comment (phretor) [reply scapy/contrib/rtps/common_types.py:200]: I'm removing this and dependent classes as they're not used anymore.

**Review by phretor — COMMENTED** [review]
Commit https://github.com/secdev/scapy/pull/3403/commits/6a985e12f6c0927be95ed730a877970d21b20fca should address the requested changes.

**Review by guedou — APPROVED** [review]
Thanks! LGTM.

@gpotter2 what is your call?

File: scapy/contrib/rtps/rtps.py
Comment (gpotter2) [reply scapy/contrib/rtps/rtps.py:401]: This was never addressed?

File: test/contrib/rtps/rtps.uts
Comment (gpotter2) [reply test/contrib/rtps/rtps.uts:53]: You are quoting a dissection test, a better build test would be for instance:
```
pkt2 = RTPS()/RTPSMessage(submessages=[
    RTPSSubMessage_HEARTBEAT(),
    ....
])
assert bytes(pkt2) == b"......"
```

**Review by gpotter2 — CHANGES_REQUESTED** [review]

**Comment by gpotter2** [comment]
One of the comments I left earlier was also not closed but not addressed.

File: test/contrib/rtps/rtps.uts
Comment (phretor) [reply test/contrib/rtps/rtps.uts:53]: OK. Fixed.

File: scapy/contrib/rtps/rtps.py
Comment (phretor) [reply scapy/contrib/rtps/rtps.py:401]: I think the right thing to do would be to use one `IntField` (4 bytes), wrapped by `EField` for endianess management.

```
    0...2...........7...............15.............23...............31
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |   HEARTBEAT   |     flags     |      octetsToNextHeader       |
    +---------------+---------------+---------------+---------------+
    | EntityId readerEntityId                                       |
    +---------------+---------------+---------------+---------------+
    | EntityId writerEntityId                                       |
    +---------------+---------------+---------------+---------------+
    |                                                               |
    + SequenceNumber firstAvailableSeqNumber                        +
    |                                                               |
    +---------------+---------------+---------------+---------------+
    |                                                               |
    + SequenceNumber lastSeqNumber                                  +
    |                                                               |
    +---------------+---------------+---------------+---------------+
    | Counter count                                                 |
    +---------------+---------------+---------------+---------------+
```

Comment (phretor) [reply scapy/contrib/rtps/rtps.py:401]: BTW, @gpotter2 I think there's a discrepancy in the specs. Not sure if @vmayoral has some ideas here.

image

It's actually 4 bytes, where only the low part matters. Is there a Scapy `Field` subclass to handle these cases?

**Comment by phretor** [comment]
> One of the comments I left earlier was also not closed but not addressed. Apart from those 2, LGTM.

See above. Thanks.

**Comment by phretor** [comment]
> One of the comments I left earlier was also not closed but not addressed. Apart from those 2, LGTM.

See above. Thanks.

> > One of the comments I left earlier was also not closed but not addressed. Apart from those 2, LGTM.
> 
> See above. Thanks.

Gentle ping @gpotter2 - I fixed/commented on your requested changes.

**Comment by vmayoral** [comment]
Thanks for the changes @phretor! 

FYI, we've got another contribution queued up dissecting another robotics protocol ([TCPROS](https://github.com/vmayoral/scapy/commit/de6c0b5f4a825870aa284fd072b93fb02a6e3546)). Once all requirements are met in here, we should try and ship that other one as well.

File: scapy/contrib/rtps/pid_types.py
Comment (gpotter2) [inline scapy/contrib/rtps/pid_types.py:21]: This shouldn't be the same for all files. This is how it appears in `explore()`
You can set `scapy.contrib.status = ignore` to have it not appear and drop the other two fields if you want. Same elsewhere

**Review by gpotter2 — DISMISSED** [review]
LGTM except a minor thing i missed earlier

File: scapy/contrib/rtps/pid_types.py
Comment (vmayoral) [reply scapy/contrib/rtps/pid_types.py:21]: @phretor this looks like a tiny change, would you like me to send a PR?

Comment (phretor) [reply scapy/contrib/rtps/pid_types.py:21]: @vmayoral yes please.

Comment (vmayoral) [reply scapy/contrib/rtps/pid_types.py:21]: @phretor @gpotter2, see https://github.com/phretor/scapy/pull/1.

Comment (gpotter2) [reply scapy/contrib/rtps/pid_types.py:21]: Pretty sure the PR is wrong (and won't pass the tests).

`explore()` is used to list all classes extending the `Packet` one. Each contrib module `must` have on of those headers if it has packets in it that you want the user to be able to find, or a header with "skip" if you don't.

See the doc: https://scapy.readthedocs.io/en/latest/build_dissect.html#add-new-protocols-to-scapy

Comment (vmayoral) [reply scapy/contrib/rtps/pid_types.py:21]: @gpotter2 @phretor I gave it a second attempt.

Comment (gpotter2) [reply scapy/contrib/rtps/pid_types.py:21]: Sure you can merge that. We'll see how it goes

Comment (vmayoral) [reply scapy/contrib/rtps/pid_types.py:21]: ping @phretor

Comment (phretor) [reply scapy/contrib/rtps/pid_types.py:21]: Sorry folks. I'm swamped by lots of things and got lost into this one. What should I do? Append `scapy.contrib.status = ignore` this to all the files?

Comment (vmayoral) [reply scapy/contrib/rtps/pid_types.py:21]: No worries @phretor, just need to merge https://github.com/phretor/scapy/pull/1. We'll see what happens.

**Comment by phretor** [comment]
Sorry folks. One rebase too much. I'm a hurry and I shouldn't work. I'll fix it when I have time. Otherwise, feel free to wipe the entire thing, copy the code from my repo, and start a fresh PR. I don't care too much about credits at this point, as long as this painful journey comes to an end.

**Comment by gpotter2** [comment]
No worries, thanks for the work !
I'll use my secret maintainer powers to rebase the PR and merge it ASAP :-)

**Comment by gpotter2** [comment]
Thanks a lot for the great PR and sorry for the time it took. I hope we didn't disgust you from contributing in scapy :/

**Pull request merged**
